app/Community/Services/SubscriptionService.php

```diff
@@ -57,58 +62,205 @@ public function isSubscribed(User $user, SubscriptionSubjectType $subjectType, i
+            }
+
+            // get explicit subscriptions (including explicitly unsubscribed so we don't fetch their implicit subscription)
+            $subscriptions = Subscription::query()
+                ->where('user_id', $user->id)
+                ->where('subject_type', $subjectType)
+                ->get();
+
+            $explicitSubscriptionIds = $subscriptions->filter(fn ($s) => $s->state === true)->pluck('subject_id')->toArray();
+
+            $implicitSubscriptionIds = $handler->getImplicitSubscriptionQuery(null, $user->id, null)
+                ->when(!empty($subscriptions), function ($query) use ($subscriptions) {
+                    $query->whereNotIn('subject_id', $subscriptions->pluck('subject_id'));
+                })
+                ->get()
+                ->pluck('subject_id')
+                ->toArray();
+
+            return array_merge($explicitSubscriptionIds, $implicitSubscriptionIds);
+       });
+    }
+
+    private function getHandler(SubscriptionSubjectType $subjectType): BaseSubscriptionHandler
     {
         return match ($subjectType) {
-            SubscriptionSubjectType::Achievement => $this->getImplicitAchievementCommentSubscriptionQuery($subjectId, $forUserId, $ignoreUserIds),
-            SubscriptionSubjectType::AchievementTicket => $this->getImplicitTicketSubscriptionQuery($subjectId, $forUserId, $ignoreUserIds),
-            SubscriptionSubjectType::ForumTopic => $this->getImplicitForumTopicSubscriptionQuery($subjectId, $forUserId, $ignoreUserIds),
-            SubscriptionSubjectType::GameAchievements => $this->getNoImplicitSubscriptionQuery(),
-            SubscriptionSubjectType::GameTickets => $this->getNoImplicitSubscriptionQuery(),
-            SubscriptionSubjectType::GameWall => $this->getImplicitCommentSubscriptionQuery(ArticleType::Game, $subjectId, $forUserId, $ignoreUserIds),
-            SubscriptionSubjectType::Leaderboard => $this->getImplicitCommentSubscriptionQuery(ArticleType::Leaderboard, $subjectId, $forUserId, $ignoreUserIds),
-            SubscriptionSubjectType::UserWall => $this->getImplicitCommentSubscriptionQuery(ArticleType::User, $subjectId, $forUserId, $ignoreUserIds),
+            SubscriptionSubjectType::Achievement => new AchievementWallSubscriptionHandler(),
+            SubscriptionSubjectType::AchievementTicket => new AchievementTicketSubscriptionHandler(),
+            SubscriptionSubjectType::ForumTopic => new ForumTopicSubscriptionHandler(),
+            SubscriptionSubjectType::GameAchievements => new GameAchievementsSubscriptionHandler(),
+            SubscriptionSubjectType::GameTickets => new GameTicketsSubscriptionHandler(),
+            SubscriptionSubjectType::GameWall => new GameWallSubscriptionHandler(),
+            SubscriptionSubjectType::Leaderboard => new LeaderboardWallSubscriptionHandler(),
+            SubscriptionSubjectType::UserWall => new UserWallSubscriptionHandler(),
         };
     }
+}
 
+abstract class BaseSubscriptionHandler
```

Comment:
I wanted these to be nested classes, but PHP doesn't support that. If they should be moved to a subfolder or something else, let me know.

They just group similar functions together by `SubscriptionSubjectType` instead of having multiple `match` statements.